host/ecallids.c

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#include "ecallids.h"
+#include <openenclave/internal/raise.h>
+#include "hostthread.h"
+
+/* The maximum numer of different ecalls that an application can make is
+ * bounded. */
+static const char* _ecall_table[OE_MAX_ECALLS];
+static uint32_t _ecall_table_size;
+
+/* Mutex for assigning/looking up global ids in a thread-safe manner. */
+static oe_mutex _lock;
+
+/* Get the global ecall id from the _ecall_table. Locking must be done by
+ * caller. */
+static oe_result_t _get_global_id(const char* name, uint64_t* global_id)
+{
+    oe_result_t result = OE_NOT_FOUND;
+    if (!name || !global_id)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    /* Search for id assigned for given name. */
+    for (uint64_t i = 0; i < _ecall_table_size; i++)
+    {
+        if (strcmp(_ecall_table[i], name) == 0)
+        {
+            *global_id = i;
+            break;
+        }
+    }
+
+    /* If the name is not found, adding it to the table. */
+    if (*global_id == OE_GLOBAL_ECALL_ID_NULL)
```

Comment:
I think *global_id should be initialized to OE_GLOBAL_ECALL_ID_NULL around line 23

Reply:
global_id is defined as static variable and initialalized in the oeedger8r generated code.

Reply:
I think it's bad programming style (bug prone) to have an IN/OUT parameter when an OUT param will do and is more appropriate.

Reply:
The reason for this is that we use the same call path that takes the global id as input. The goal is setting the global id once (non constant time lookup), and use the cached value for the subsequent invocations (constant time lookup). Not sure if OUT param can achieve this goal.

Reply:
Yes, as @mingweishih described, global_id is an in/out parameter so as to cache the result of the name to global identifier lookup.

For each ecall wrapper function in the host, edger8r generates a static `global_id` variable:
```c
oe_result_t ecall_foo_wrapper(...)
{
   static uint64_t global_id = OE_GLOBAL_ECALL_ID_NULL; // for caching the look up
   ...
   OE_CHECK(oe_get_ecall_ids(enclave, "foo", &global_id, &local_id); // look up global id from name and cache it.
   ...
   dispatch-ecall(enclave, local_id, args);
}
```

Alternatively, we could do this:
edger8r generates an array of global_id caching variables per enclave (user edl):

```
static uint64_t _global_ids[NUM_ECALLS_IN_ENCLAVE]; // 1 element per ecall.
```

`oe_register_ecalls` (which is called during enclave construction) is passed this array (as an out parameter) and it initializes it.

```
oe_result_t oe_register_ecalls(
    oe_enclave_t* enclave,
    const oe_ecall_info_t* ecall_info_table,
    uint32_t num_ecalls,
    uint64_t* global_ids) // out parameter
{
   ...
   for (uint32_t i=0; i < num_ecalls; ++i)
       oe_get_ecall_id(ecall_info_table[i].name, &global_ids[i]); // populate cache
   ...
}
```
Since global id has already been populated during `oe_register_ecalls`, ecalls can be dispatched as:
```c
oe_result_t ecall_foo_wrapper(...)
{
   static uint64_t global_id = OE_GLOBAL_ECALL_ID_NULL; // for caching the look up
   ...
   OE_CHECK(oe_get_ecall_function_id(
                  enclave, 
                  "foo", 
                  _global_id[fcn_idx], // pass in the global id for ecall to get the local id
                  &local_id); 
   ...
   dispatch-ecall(enclave, local_id, args);
}
```

Reply:
@dthaler @mingweishih 
Above, I've proposed an alternate strategy of initializing global ids.
I'm fine with the current approach in this PR or the new proposal.
Let me know what you think.

Reply:
@anakrish Do you think there is other benefit of adopting the alternative approach (maybe related to sharing tables among the enclaves with the same instances)? I'm just a bit concerned because doing so requires changes to oeedger8r.

Reply:
@mingweishih Yes, the alternative approach localizes all initialization to `oe_register_ecalls` function, and in future could help with sharing tables between enclaves created from the same edl/binary.

However, given the scope of the current PR, sharing tables is not a must have, and I'd be OK with either approach.

The current implementation does have the advantage that it doesn't need any edger8r changes.

Reply:
Thanks for the clarification. In this case, I will prefer the current approach given the concern of IN/OUT would go away when we implement the table sharing.